app/post_metadata.go

```diff
@@ -0,0 +1,418 @@
+		Posts: make(map[string]*model.Post, numPosts),
+		Order: originalList.Order, // Note that this uses the original Order array, so it isn't a deep copy
+	}
+
+	posts := make(chan *model.Post, numPosts)
+
+	for _, originalPost := range originalList.Posts {
+		go func(originalPost *model.Post) {
+			posts <- a.PreparePostForClient(originalPost)
+		}(originalPost)
+	}
+
+	for i := 0; i < numPosts; i++ {
+		post := <-posts
+		list.Posts[post.Id] = post
+	}
+
+	close(posts)
+
+	return list
+}
+
+func (a *App) PreparePostForClient(originalPost *model.Post) *model.Post {
+	post := originalPost.Clone()
+
+	// Proxy image links before constructing metadata so that requests go through the proxy
+	post = a.PostWithProxyAddedToImageURLs(post)
+
+	if post.Metadata == nil {
+		post.Metadata = &model.PostMetadata{}
+
+		// Emojis and reaction counts
+		if emojis, reactions, err := a.getEmojisAndReactionsForPost(post); err != nil {
+			mlog.Warn("Failed to get emojis and reactions for a post", mlog.String("post_id", post.Id), mlog.Any("err", err))
+		} else {
+			post.Metadata.Emojis = emojis
+			post.Metadata.Reactions = reactions
+		}
+
+		// Files
+		if fileInfos, err := a.getFileMetadataForPost(post); err != nil {
+			mlog.Warn("Failed to get files for a post", mlog.String("post_id", post.Id), mlog.Any("err", err))
+		} else {
+			post.Metadata.Files = fileInfos
+		}
+
+		// Embeds and image dimensions
+		firstLink, images := getFirstLinkAndImages(post.Message)
+
+		if embed, err := a.getEmbedForPost(post, firstLink); err != nil {
+			mlog.Warn("Failed to get embedded content for a post", mlog.String("post_id", post.Id), mlog.Any("err", err))
+		} else if embed == nil {
+			post.Metadata.Embeds = []*model.PostEmbed{}
+		} else {
+			post.Metadata.Embeds = []*model.PostEmbed{embed}
+		}
+
+		post.Metadata.Images = a.getImagesForPost(post, images)
+	}
+
+	return post
+}
+
+func (a *App) getFileMetadataForPost(post *model.Post) ([]*model.FileInfo, *model.AppError) {
+	if len(post.FileIds) == 0 { // This field is deprecated, but still use it for now to avoid unnecessary database hits
+		return nil, nil
+	}
+
+	return a.GetFileInfosForPost(post.Id, false)
+}
+
+func (a *App) getEmojisAndReactionsForPost(post *model.Post) ([]*model.Emoji, []*model.Reaction, *model.AppError) {
+	var reactions []*model.Reaction
+	if post.HasReactions { // This field is deprecated, but still use it for now to avoid unnecessary database hits
+		var err *model.AppError
+		reactions, err = a.GetReactionsForPost(post.Id)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	emojis, err := a.getCustomEmojisForPost(post, reactions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return emojis, reactions, nil
+}
+
+func (a *App) getEmbedForPost(post *model.Post, firstLink string) (*model.PostEmbed, error) {
+	if _, ok := post.Props["attachments"]; ok {
+		return &model.PostEmbed{
+			Type: model.POST_EMBED_MESSAGE_ATTACHMENT,
+		}, nil
+	}
+
+	if firstLink != "" {
```

Comment:
I was leaning towards having the only nil return at the end of the function, but since we're having an early return if `getLinkMetadata` returns an error, it makes sense to also make this return early as well